atl-checker/src/algorithms/certain_zero/search_strategy/linear_programming_search.rs

```diff
@@ -0,0 +1,224 @@
+}
+
+impl SearchStrategy<AtlVertex> for LinearProgrammingSearch {
+    /// Simply returns the edge with highest priority (i.e lowest distance)
+    fn next(&mut self) -> Option<Edge<AtlVertex>> {
+        self.queue.pop().map(|(edge, _)| edge)
+    }
+
+    /// Takes a Vec of Edges holding ATLVertices, and puts these in the queue,
+    /// based on the calculated distance from its state to acceptance region from formula
+    fn queue_new_edges(&mut self, edges: Vec<Edge<AtlVertex>>) {
+        for edge in edges {
+            let distance = self.get_distance_of_edge(&edge);
+
+            // Add edge and distance to queue
+            if let Some(dist) = distance {
+                self.queue.push(edge, -dist);
+            } else {
+                // Todo what should default value be, if cannot be calculated? For now: first priority
+                self.queue.push(edge, 0);
+            }
+        }
+    }
+}
+
+impl LinearProgrammingSearch {
+    /// Finds the distance of an edge using linear programming
+    fn get_distance_of_edge(&mut self, edge: &Edge<AtlVertex>) -> Option<i32> {
+        let source = edge.source();
+
+        // If we have seen this source before, get the result instantly
+        if let Some(sol) = self.result_cache.get(&source) {
+            return Some(sol.1);
+        }
+
+        // If we have not seen this formula before, calculate constrained phi
+        if !self.phi_cache.contains_key(&source.formula()) {
+            // Convert the formula to a structure of constraints
+            let lcp = self.phi_mapper.map(&self.game, &*source.formula());
+            self.phi_cache.insert(source.formula(), lcp);
+        }
+
+        // Get constraints of this phi
+        let constrained_phi = self.phi_cache.get(&source.formula()).unwrap();
+        // Get current state in vertex
+        let state = self.game.state_from_index(source.state());
+
+        // Use linear programming to find closest state that satisfies the constraints
+        if let Some(sol) = self.get_optimal_distance(constrained_phi, &state) {
+            let dist = sol.1;
+            // Cache the resulting solution
+            self.result_cache.insert(source.clone(), sol);
+            return Some(dist);
+        }
+        // If we could not find a distance, return None
+        None
+    }
+
+    fn get_optimal_distance(
+        &self,
+        constrained_phi: &LinearConstrainedPhi,
+        state: &State,
+    ) -> Option<(State, i32)> {
+        // The LinearConstrainedPhi contains multiple variants of the linear problem due to the ORs.
+        // So we iterate through them and find the best result
+        let mut best: Option<(State, i32)> = None;
+        for constraints in all_variants(constrained_phi) {
+            // We keep track of the symbols/variables encountered
+            let mut symbol_vars: HashMap<SymbolIdentifier, Variable> = HashMap::new();
+            let mut problem = Problem::new(OptimizationDirection::Minimize);
```

Comment:
Also, why Minimize? (not because I disagree, just to hear your thoughts)

Reply:
We want to find the closest satisifying state, aka minimum distance.

Reply:
The reason I ask, is because I wondered myself what to do here. Because, if we want to measure distance to the "acceptance border", or the point on satisfying/not satisfying, right? In that case, 
![image](https://user-images.githubusercontent.com/33687173/118393096-82e15d80-b63d-11eb-984f-0887d37a5654.png)
If we minimize in this scenario, where state is at origin, green area is acceptance, then will we get a distance of 0, because the linear programming will try to minimize all variables in the solution. Same sort of problem in the maximize direction. It is only if we are "below" the constraint and maximize in this situation we get the distance to the constraint. 
And in the same way, it is only if we are "above" the constraint and minimize, we get the distance to the constraint. 
Or am I understanding it wrongly?

Reply:
With linear programming we want to find the minimum distance to a state that satisifies the formula, not the distance to the acceptance border. The other heuristic we made searches towards acceptrance borders, which is why it is so unique.